File: Helpers/src/main/java/com/dg/helpers/ObjectHelper.java
package com.dg.helpers;

import org.json.JSONObject;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 */
public class ObjectHelper
{
    private static Object JSONNULL = JSONObject.NULL;

    /**
     * Checks for equality between two objects.
     * * Normalizes {@code JSONObject.NULL} to null
     * * Successfully compares Float with Double
     * * Successfully compares Integer with Long
     * @param left
     * @param right
     * @return true if equal
     */
    public static boolean equals(Object left, Object right)
    {
        if (left == JSONNULL)
        {
            left = null;
        }

        if (right == JSONNULL)
        {
            right = null;
        }

        if (left == null && right != null)
        {
            return false;
        }

        if (right == null && left != null)
        {
            return false;
        }

        if (left == null)
        {
            return true;
        }

        if (left instanceof Float && right instanceof Double)
        {
            return right.equals((double)(float)(Float)left);
        }

        if (right instanceof Float && left instanceof Double)
        {
            return left.equals((double)(float)(Float)right);
        }

        if (left instanceof Integer && right instanceof Long)
        {
            return right.equals((long)(int)(Integer)left);
        }

        if (right instanceof Integer && left instanceof Long)
        {
            return left.equals((long)(int)(Integer)right);
        }

        return left.equals(right);
    }
}

File: Helpers/src/main/java/com/dg/helpers/ObjectHelper.kt
package com.dg.helpers

import org.json.JSONObject

@Suppress("unused")
object ObjectHelper
{
    private val JSON_NULL = JSONObject.NULL

    /**
     * Checks for equality between two objects.
     * * Normalizes `JSONObject.NULL` to null
     * * Successfully compares Float with Double
     * * Successfully compares Integer with Long
     * @param left
     * @param right
     * @return true if equal
     */
    fun equals(left: Any?, right: Any?): Boolean
    {
        var left = left
        var right = right
        if (left === JSON_NULL)
        {
            left = null
        }

        if (right === JSON_NULL)
        {
            right = null
        }

        if (left == null && right != null)
        {
            return false
        }

        if (right == null && left != null)
        {
            return false
        }

        if (left == null)
        {
            return true
        }

        if (left is Float && right is Double)
        {
            return right == left.toDouble()
        }

        if (right is Float && left is Double)
        {
            return left == right.toDouble()
        }

        if (left is Int && right is Long)
        {
            return right == left.toLong()
        }

        return if (right is Int && left is Long)
        {
            left == right.toLong()
        }
        else left == right
    }
}
